examples/simple/simple.cpp

```diff
@@ -0,0 +1,208 @@
+
+    // load data from cpu memory to backend buffer
+    if(ggml_backend_buffer_is_host(model.buffer)) {
+        memcpy(model.a->data, a, ggml_nbytes(model.a));
+        memcpy(model.b->data, b, ggml_nbytes(model.b));
+    } else {
+        // CUDA backend requires copy the data directly to device
+        ggml_backend_tensor_set(model.a, a, 0, ggml_nbytes(model.a));
+        ggml_backend_tensor_set(model.b, b, 0, ggml_nbytes(model.b));
+    }
+}
+
+// build the compute graph to perform a matrix multiplication
+
+struct ggml_cgraph * build_graph(const simple_model& model, struct ggml_allocr * allocr) {
+    static size_t buf_size = ggml_tensor_overhead()*GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_allocr_alloc_graph()
+    };
+
+    // create a temporally context to build the graph
+    struct ggml_context * ctx0 = ggml_init(params0);
+
+    struct ggml_cgraph  * gf = ggml_new_graph(ctx0);
+
+    // result = a*b^T
+    struct ggml_tensor* result = ggml_mul_mat(ctx0, model.a, model.b);
+
+    ggml_build_forward_expand(gf, result);
+
+    // delete the temporally context used to build the graph
+    ggml_free(ctx0);
+    return gf;
+}
+
+// compute with backend
+struct ggml_tensor* compute(const simple_model & model, struct ggml_allocr * allocr) {
+    // reset the allocator to free all the memory allocated during the previous inference
+    ggml_allocr_reset(allocr);
+
+    struct ggml_cgraph * gf = build_graph(model, allocr);
+
+    // allocate tensors
+    ggml_allocr_alloc_graph(allocr, gf);
+
+    int n_threads = 1; // number of threads to perform some operations with multi-threading
+
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, n_threads);
+    }
+
+#ifdef GGML_USE_METAL
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_cb(model.backend, n_threads);
+    }
+#endif
+
+    ggml_backend_graph_compute(model.backend, gf);
+
+    // in this case, the output tensor is the last one in the graph
+    return gf->nodes[gf->n_nodes - 1];
+}
+
+int main(void)
+{
+    ggml_time_init();
+
+    // initialize data of matrices to perfom matrix multiplication
+
+    float matrix_A[4] = {
+        2, 8,
+        5, 1
+    };
+
+    // transpose of [10, 5, 9, 9]
+    float matrix_B[4] = {
+        10, 9,
+        5, 9
+    };
+
+    simple_model model;
+    load_model(model, matrix_A, matrix_B);
+
+    // calculate the temporaly memory required to compute
+    ggml_backend_buffer_t buf_compute;
+    struct ggml_allocr * allocr = NULL;
+
+    {
+        size_t align = ggml_backend_get_alignment(model.backend);
+        allocr = ggml_allocr_new_measure(align);
```

Comment:
It's better to use `ggml_allocr_new_measure_from_backend`, then the alignment and padding required by the backend is correctly taken into account.